File: sorting/C_Cpp/merge_k_sorted_lists.c
#include <stdlib.h>

typedef struct ListNode {
    int val;
    struct ListNode* next;
} ListNode;

struct ListNode* merge(struct ListNode* list1, struct ListNode* list2);

struct ListNode* mergeKLists(struct ListNode** lists, int listsSize) {

    int n = listsSize;
    if (!lists || n == 0) {
        return NULL;
    }

    while (n > 1) {
        ListNode** temp = (ListNode**)malloc(sizeof(ListNode*) * (n / 2 + n % 2));
        int k = 0;
        for (int i = 0; i < n; i += 2) {
            ListNode* l1 = lists[i];
            ListNode* l2 = (i+1 < n) ? lists[i+1] : NULL;
            temp[k] = merge(l1,l2);
            k++;
        }
        lists = temp;
    }
    return lists[0];
}

struct ListNode* merge(struct ListNode* list1, struct ListNode* list2) {

    struct ListNode* dummy = malloc(sizeof(struct ListNode));
    struct ListNode* tail = dummy;

    while(list1 && list2) {
        if (list1->val < list2->val) {
            tail->next = list1;
            list1 = list1->next;
        }
        else {
            tail->next = list2;
            list2 = list2->next;
        }
        tail = tail->next;
    }
    if (list1) {
        tail->next = list1;
    }
    else if (list2) {
        tail->next = list2;
    }

    return dummy->next;
}

File: sorting/Python/merge_k_sorted_lists.py
class ListNode(self, val=0, next=None):
    self.val = val
    self.next = next


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if not lists or len(lists) == 0:
            return None

        while len(lists) > 1:
            temp = []
            for i in range(0, len(lists), 2):
                l1 = lists[i]
                l2 = lists[i+1] if i+1 < len(lists) else None
                temp.append(self.merge(l1, l2))
            lists = temp
        return lists[0]


    def merge(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode()
        tail = dummy

        while l1 or l2:
            if l1 and ((not l2) or l1.val < l2.val):
                tail.next = l1
                l1 = l1.next

            else:
                tail.next = l2
                l2 = l2.next
            tail = tail.next

        return dummy.next

